### ISSUE TITLE
Temp file cannot be accessed because it is being used by another process, VS becomes unresponsive

### ISSUE THREAD

**Issue by torronen** [dotnet/machinelearning-modelbuilder#1535]
**System Information (please complete the following information):**
- Model Builder Version (available in Manage Extensions dialog): 16.6.1.2131904
- Visual Studio Version: 16.10.2

**Describe the bug**
The error occurred after opening the Advanced Data Settings window and waiting (writing another bug description).

Advanced Data Settings window showed "The process cannot access the file ... {tempFile} ... " after opening. 
The error may be rare, but after clicking "close" the Advanced Data Settings window become unresponsive and can not be closed or moved. It can not even be closed from the Start menu thumbnail. 

**To Reproduce**
Not easily reproducible.

**Expected behavior**
Model builder should not need locked access to the temp file at this time, but this is the first time I encountered it.
Bigger issue is making visual studio and model builder modal "un-closable" (otherwise it seems still responsive based on blinking window borders when trying to click - but nothing happens)

**Screenshots**
![image](https://user-images.githubusercontent.com/26261427/123645700-389fee80-d82f-11eb-9ea9-b00145dfec66.png)
![image](https://user-images.githubusercontent.com/26261427/123646350-cc71ba80-d82f-11eb-8bb3-1c71829351f1.png)

**Additional context**
I did not identify what was using the file, but antivirus was doing system scan meanwhile.

**Comment by LittleLittleCloud** [comment]
Did you have another model builder instance start training at the same time when this error thrown? That file is the temp dataset created by model builder automl service.

The temp dataset's name is generated from a random seed, which the random seed is initialized using the following way.
![image](https://user-images.githubusercontent.com/16876986/123865991-680f3380-d8e1-11eb-8ed5-6ab5f2ddb5e3.png)

This is a potential cause for this `Temp file can't be accessed` error as the likelihood of two integers from `random.next()` to be the same is not too small to be ignored. So it might be the new model builder happens to generate an identical random dataset name which is currently used by an on-going training from another model builder instance

**Comment by torronen** [comment]
I think it was at the time issue https://github.com/dotnet/machinelearning-modelbuilder/issues/1533 happened, so yes, there was another training happening in the background, I believe, even if I could not access it. I was writing about that issue, and noticed this temp file error when coming back to Visual Studio.

**Comment by LittleLittleCloud** [comment]
Thanks @torronen , we will start working on a fix to make the temp dataset's name more "random". In the meanwhile, a simple work-around can be re-launching training (if possible). We will look into the frozen VS issue as well, but it's very like that the two issues are unrelated.

**Comment by torronen** [comment]
Sorry, but I can no longer verify if the Kaspersky antivirus was scanning the specific folder at that time, but it did not warn or detect any items so it should not have changed anything. However, I think the chance of antivirus scanning this file has a low probability because scanning a single file is quick and it was doing full system scan with lots of files. I can try to reproduce it on next run with antivirus but probably will not be succesful.

 This is probably a low impact bug for most, I only reported it as I was already on a good run. I've seen this issue on only once for 100+ training runs, and I did not hear about this issue from my team either.

**Comment by LittleLittleCloud** [comment]
And did you remember what's the last trainer before crash? Is that all lgbm?

**Comment by LittleLittleCloud** [comment]
K, In that case, I'm going to close this issue, please reopen it and let us know if you hit another similar error

**Comment by torronen** [comment]
Issue was found again during a multiday run. The Log file indicated in the error does not exist. There are .zip files (models probably) and a binary file in the directory. It seems VS also closed the Output window of Machine Learning so I can not see which one was the best model. 

In addition to this and https://github.com/dotnet/machinelearning-modelbuilder/issues/1531 I think we should expect that long training runs may often fail, becuase 1531 and 1535 may be "hiding" other issues which rarely occur now because either of these may crash the progress. Could I find which one was the best model from some of the log files / is it possible to recover?
The alternative is to try to guess when is best time to cancel and hope it does not crash before.

This machine was Windows Server 2019. There should not have been any simultaneous training, at least not intended.


<img width="412" alt="Capture" src="https://user-images.githubusercontent.com/26261427/125162229-62420980-e18f-11eb-9d50-179b9c5d7c5d.PNG">
<img width="892" alt="Capture2" src="https://user-images.githubusercontent.com/26261427/125162230-62daa000-e18f-11eb-8c6c-e84e95f0bbf2.PNG">
<img width="500" alt="files" src="https://user-images.githubusercontent.com/26261427/125162231-63733680-e18f-11eb-8476-dc623204f8b8.PNG">


BTW, I noted slowness of Windows UI even if I could see almost no CPU or RAM usage in the Resource monitors until I closed Visual Studio. Could be a coincidence for some other reason.

**Comment by torronen** [comment]
If "recoverable models" need a new log file, it maybe is needed also for https://github.com/dotnet/machinelearning-modelbuilder/issues/1571 as crash may happen in Azure, too. Users may be less happy if they pay for a high-end VM and it ends up crashing without any output. One more reason  for "recover latest" is that new algorithms might be introduced later, and as they are included in Model Builder, they may again start causing crashes even if the current ones would be fully fixed.

**Comment by torronen** [comment]
It seems .mbconfig Train tab retains the data for the best found model, even if the Model.zip was not copied to solution ( https://github.com/dotnet/machinelearning-modelbuilder/issues/1582 )

**Comment by torronen** [comment]
Added separate issue about "Recoverable models" https://github.com/dotnet/machinelearning-modelbuilder/issues/1584

**Comment by torronen** [comment]
This server did not have realtime antivirus, not even Windows Defender.

**Comment by LittleLittleCloud** [comment]
Hi @torronen 

> Could I find which one was the best model from some of the log files / is it possible to recover?

You can find all the output in log, Here's an example for one iteration. Using that iteration number, you should be able to find the best model in AutoML-NNI folder. You can find the path to the log on the very top of output window, or in %temp%MLVSTools\logs, where all logs will be retained.

![image](https://user-images.githubusercontent.com/16876986/125347220-83316880-e30f-11eb-8b6a-3fc4026df811.png)

> Issue was found again during a multiday run. The Log file indicated in the error does not exist. 
It's a known bug and you should still be able to find all retained logs in %temp%MLVSTools\logs, can you unzip all logs and upload here, thanks!

**Comment by torronen** [comment]
Here is log file from latest from %temp%MLVSTools\logs  Only one file
[939190b5-6b26-450c-9dcf-7a0de7351539.txt](https://github.com/dotnet/machinelearning-modelbuilder/files/6847655/939190b5-6b26-450c-9dcf-7a0de7351539.txt)

Error:
`   at System.IO.__Error.WinIOError(Int32 errorCode, String maybeFullPath)
   at System.IO.FileStream.Init(String path, FileMode mode, FileAccess access, Int32 rights, Boolean useRights, FileShare share, Int32 bufferSize, FileOptions options, SECURITY_ATTRIBUTES secAttrs, String msgPath, Boolean bFromProxy, Boolean useLongPath, Boolean checkHost)
   at System.IO.FileStream..ctor(String path, FileMode mode, FileAccess access, FileShare share, Int32 bufferSize, FileOptions options, String msgPath, Boolean bFromProxy)
   at System.IO.FileStream..ctor(String path, FileMode mode)
   at Microsoft.ML.ModelBuilder.AutoMLService.Proposer.ResampleStrategyProposer.Initialize(ProposerSharedContext context, String outputFolder) in /_/src/Microsoft.ML.ModelBuilder.AutoMLService/Proposer/ResampleStrategyProposer.cs:line 83
   at Microsoft.ML.ModelBuilder.AutoMLService.Proposer.ResampleStrategyProposer.Propose(ProposerSharedContext context, TrialParameter trialParameter) in /_/src/Microsoft.ML.ModelBuilder.AutoMLService/Proposer/ResampleStrategyProposer.cs:line 48
   at Microsoft.ML.ModelBuilder.AutoMLService.Proposer.Controller.Propose() in /_/src/Microsoft.ML.ModelBuilder.AutoMLService/Proposer/Controller.cs:line 63
   at Microsoft.ML.ModelBuilder.AutoMLService.Experiments.LocalAutoMLExperiment.<ExecuteAsync>d__15.MoveNext() in /_/src/Microsoft.ML.ModelBuilder.AutoMLService/Experiments/LocalAutoMLExperiment.cs:line 130
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.ML.ModelBuilder.AutoMLEngine.<StartTrainingAsync>d__23.MoveNext() in /_/src/Microsoft.ML.ModelBuilder.AutoMLService/AutoMLEngineService/AutoMLEngine.cs:line 153`

![image](https://user-images.githubusercontent.com/26261427/126312856-ae03b269-6150-477f-a96a-14ef8e174d80.png)


File %temp%\2\AutoML-NNI\L7XF3W does not exist

**Comment by LittleLittleCloud** [comment]
@vzhuqin 

This is hard to reproduce, feel free to skip it for weekly test pass.

**Comment by torronen** [comment]
To reproduce. may need to have a VM running for 1 or more days. I have multiple machines running, about maybe about 10% failure rate per day due to this exception with my data (1-2 gb datasets) but varies a lot per day. Happy to update here if it is no longer occuring for me after there should be a fix for it.

**Comment by torronen** [comment]
Possible hint: It would seem that the issue happens after a long training run more often than otherwise. Is it possible operating system does something meanwhile, or maybe VS "switches" to another process while the lock was created from another one. Or, maybe experiment 575 has some finalization tasks remaining which normally take milliseconds but with a large model take much longer, and then it tries to start experiment 576. Just guessing, fully :D 

File SFUKJW exists
![image](https://user-images.githubusercontent.com/26261427/129842825-29f5e37e-9f70-4786-8044-0cc7b6663b7e.png)


[1d667f82-5c69-4a9a-9c67-25cdfa5f7870.txt](https://github.com/dotnet/machinelearning-modelbuilder/files/7004728/1d667f82-5c69-4a9a-9c67-25cdfa5f7870.txt)

![image](https://user-images.githubusercontent.com/26261427/129842410-46853ca3-33e3-495c-933f-4102ce4eaad0.png)

**Comment by torronen** [comment]
The zip filenames in the folder and the iteration run numbers do not much.
![image](https://user-images.githubusercontent.com/26261427/129843317-0435164e-96f7-4242-b2ec-d9cb18093d73.png)

It seems this became an extreme example. Maybe some version could have an advanced setting to set the maximum size of trees. 6Gb zipped model for a non-zipped 62Gb CSV seems too big.

**Comment by LittleLittleCloud** [comment]
@torronen 
Thanks for the provided info, the 1-line log file is a bug which we already fixed and will soon been released. And we will look into the mismatch model name.

**Comment by torronen** [comment]
I've recently experienced this issue less often, so I am thinking it might be something to what I have been doing. Do you know if other people or testing are experiencing same issue? 

If others are not, then one more guess is it could be related to opening and setting up another .mbconfig (but not run training yet)

**Comment by torronen** [comment]
@LittleLittleCloud It might be a coincidence, but in this case run was okay for a long time, until I started opening other .mbconfig files and then the error happened. 
![image](https://user-images.githubusercontent.com/26261427/130354613-04f3514a-1f51-4ae0-bd78-cbbe5472bc7d.png)

2021-08-22 04:44:29.9509 DEBUG The process cannot access the file 'C:\Users\Administrator\AppData\Local\Temp\2\AutoML-NNI\MGWQ3D' because it is being used by another process.
   at System.IO.__Error.WinIOError(Int32 errorCode, String maybeFullPath)
   at System.IO.FileStream.Init(String path, FileMode mode, FileAccess access, Int32 rights, Boolean useRights, FileShare share, Int32 bufferSize, FileOptions options, SECURITY_ATTRIBUTES secAttrs, String msgPath, Boolean bFromProxy, Boolean useLongPath, Boolean checkHost)
   at System.IO.FileStream..ctor(String path, FileMode mode, FileAccess access, FileShare share, Int32 bufferSize, FileOptions options, String msgPath, Boolean bFromProxy)
   at System.IO.FileStream..ctor(String path, FileMode mode)
   at Microsoft.ML.ModelBuilder.AutoMLService.Proposer.ResampleStrategyProposer.Initialize(ProposerSharedContext context, String outputFolder) in /_/src/Microsoft.ML.ModelBuilder.AutoMLService/Proposer/ResampleStrategyProposer.cs:line 83
   at Microsoft.ML.ModelBuilder.AutoMLService.Proposer.ResampleStrategyProposer.Propose(ProposerSharedContext context, TrialParameter trialParameter) in /_/src/Microsoft.ML.ModelBuilder.AutoMLService/Proposer/ResampleStrategyProposer.cs:line 48
   at Microsoft.ML.ModelBuilder.AutoMLService.Proposer.Controller.Propose() in /_/src/Microsoft.ML.ModelBuilder.AutoMLService/Proposer/Controller.cs:line 63
   at Microsoft.ML.ModelBuilder.AutoMLService.Experiments.LocalAutoMLExperiment.<ExecuteAsync>d__15.MoveNext() in /_/src/Microsoft.ML.ModelBuilder.AutoMLService/Experiments/LocalAutoMLExperiment.cs:line 130
--- End of stack trace from previous location where exception was thrown ---
   at System.Runtime.ExceptionServices.ExceptionDispatchInfo.Throw()
   at System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
   at Microsoft.ML.ModelBuilder.AutoMLEngine.<StartTrainingAsync>d__23.MoveNext() in /_/src/Microsoft.ML.ModelBuilder.AutoMLService/AutoMLEngineService/AutoMLEngine.cs:line 153 (Microsoft.ML.ModelBuilder.Utils.Logger.Debug)
   
    Very annoying problem :(

**Comment by luisquintanilla** [comment]
Related to #1645. Issue should be fixed now. @torronen feel free to open a new issue if you're still having problems.
